### PR TITLE
Client-side file validation in upload for gzipped files (SCP-4160)

### PR DESCRIPTION
**Pull request by eweitz**
This adds partial CSFV for files compressed via gzip.  It drastically improves time-to-fail on a large portion of uploaded files.

Previously, gzipped files had superficial validation, checking for a proper first byte and file extension.  Now, all CSFV rules are applied for up to 500 lines of an uploaded gzipped file, newly benefitting [~15% of uploaded files](https://mixpanel.com/s/4xDTGv).  Decompression is handled by the small but mighty [fflate](https://github.com/101arrowz/fflate) JavaScript library.

Here's how it looks:

https://user-images.githubusercontent.com/1334561/158390985-02d50413-2fce-4247-b913-8275f5de5a36.mov

Also included is preliminary refactoring for live reload using Vite.  This builds on #1397.  Live reload works for React components like `FileUploadControl.js`, but is still a work-in-progress for libraries like `validate-file.js`, `io.js`, etc.

To test:
* Go to upload page
* Upload an invalid gzipped file, e.g. [Dense_matrix_204_MB_bad_GENE_header.txt.gz](https://storage.googleapis.com/broad-singlecellportal-public/test/Dense_matrix_204_MB_bad_GENE_header.txt.gz)
* Confirm error message and warning message appear
* Upload a valid gzipped file, e.g. [Dense_matrix_204_MB_good.txt.gz](https://storage.googleapis.com/broad-singlecellportal-public/test/Dense_matrix_204_MB_good.txt.gz)
* Confirm error message does not appear, and warning message does appear

This satisfies SCP-4160.

### CODE REVIEW

File: app/javascript/lib/validation/chunked-line-reader.js [hunk 1]
```diff
@@ -1,5 +1,8 @@
 import { DEFAULT_CHUNK_SIZE, oneGiB, readFileBytes } from './io'
 
+/** Mitigates UI freezes caused by faux-streaming gunzi; see note in io.js */
```
Comment (devonbush) [inline app/javascript/lib/validation/chunked-line-reader.js:3]:
```suggestion
/** Mitigates UI freezes caused by faux-streaming gunzip; see note in io.js */
```

File: app/javascript/lib/validation/validate-file-content.js [hunk 2]
```diff
@@ -356,12 +356,16 @@ export async function parseFile(file, fileType, fileOptions={}, sizeProps={}) {
...
+
   const parseResult = { fileInfo, issues: [] }
 
   try {
     fileInfo.isGzipped = await validateGzipEncoding(file)
     // if the file is compressed or we can't figure out the compression, don't try to parse further
-    if (fileInfo.isGzipped || !PARSEABLE_TYPES.includes(fileType)) {
+    if (
+      !PARSEABLE_TYPES.includes(fileType) ||
+      file.size > sizeProps?.fileSizeTotal && fileInfo.isGzipped // Avoids bug in sync gzip CSFV
```
Comment (devonbush) [inline app/javascript/lib/validation/validate-file-content.js:367]: what bug? can you clarify?

**Review by devonbush — CHANGES_REQUESTED** [review]
This is good stuff--addresses a major gap in CSFV.  The one blocker I see is one reference to a bug? that needs a clearer comment.  Also, can you explain why we've set the chunk size higher for gzipped files than for regular files?  Would we get better UX performance with a smaller chunk size?

File: app/javascript/lib/validation/validate-file-content.js (hunk 2 above)
Comment (eweitz) [reply app/javascript/lib/validation/validate-file-content.js]: I've refined and extended this comment in d5b0b00.  

Thanks for prompting another look here -- I found an actual bug in this updated CSFV handling for "large" (> 50 MiB) gzipped files.  That's fixed now too.

**Review by ehanna4 — APPROVED** [review]
LGTM - tested locally and functionality worked as described

As a future thought - the upload warning/error section is getting to be a little wordy when there's warning/s and error/s and I wonder if we should consider a UX-y update to make the most important bits stand out more
Screen Shot 2022-03-16 at 4 20 09 PM

File: app/javascript/lib/validation/io.js [hunk 3]
```diff
@@ -20,8 +22,49 @@ export async function readFileBytes(file, startByte=0, chunkSize=DEFAULT_CHUNK_S
...
+  // - https://github.com/101arrowz/fflate/discussions/30
+  // - https://github.com/101arrowz/fflate/discussions/119
+  //
+  // Consider refactoring ChunkedLineReader to enable real streaming-gunzip,
+  // to mitigate UI freezes for non-tiny gzipped files.  This might
+  // make sense to do as part of a larger refactor to parallelize
+  // byte-upload and file validation.
+  const uint8Array = new Uint8Array(arrayBuffer)
+
+  // const stringContent = strFromU8(gunzipSync(uint8Array))
```
Comment (devonbush) [inline app/javascript/lib/validation/io.js:46]: did you mean to leave in this commented-out line?

File: app/javascript/lib/validation/validate-file-content.js [hunk 4]
```diff
@@ -387,14 +405,27 @@ export async function parseFile(file, fileType, fileOptions={}, sizeProps={}) {
...
         await parseFunctions[fileType](chunker, fileInfo.fileMimeType, fileOptions)
       fileInfo.linesRead = chunker.linesRead
       fileInfo.delimiter = delimiter
       fileInfo.numColumns = numColumns
       parseResult.issues = parseResult.issues.concat(issues)
+
+      if (fileInfo.isGzipped && fileInfo.linesRead === GZIP_MAX_LINES + 1) {
+        const msg =
+        'Due to this file\'s size, it will be fully validated after upload, ' +
+        'and any errors will be emailed to you.'
```
Comment (devonbush) [inline app/javascript/lib/validation/validate-file-content.js:425]: I think backticks are preferred over '+' for multi-line strings

File: app/javascript/lib/validation/validate-file-content.js [hunk 5]
```diff
@@ -356,12 +367,19 @@ export async function parseFile(file, fileType, fileOptions={}, sizeProps={}) {
...
     fileInfo.isGzipped = await validateGzipEncoding(file)
     // if the file is compressed or we can't figure out the compression, don't try to parse further
-    if (fileInfo.isGzipped || !PARSEABLE_TYPES.includes(fileType)) {
+    if (
+      !PARSEABLE_TYPES.includes(fileType) ||
+      fileInfo.isGzipped && (
+        file.size > sizeProps?.fileSizeTotal || // CSFV for gzip needs whole file
+        file.size >= MAX_GZIP_FILESIZE // See comment for MAX_GZIP_FILESIZE
+      )
+    ) {
```
Comment (devonbush) [inline app/javascript/lib/validation/validate-file-content.js:382]: An intermediate variable will help make the if block less complex to read.  
```suggestion
    const isFileFragment = file.size > sizeProps?.fileSizeTotal // likely a partial download from a GCP bucket
    if (
      !PARSEABLE_TYPES.includes(fileType) ||
      fileInfo.isGzipped && (isFileFragment || file.size >= MAX_GZIP_FILESIZE)
      // current gunzip implementation needs a whole file; see comment for MAX_GZIP_FILESIZE
    ) {
```

File: app/javascript/lib/validation/validate-file-content.js [hunk 6]
```diff
@@ -387,14 +405,27 @@ export async function parseFile(file, fileType, fileOptions={}, sizeProps={}) {
 
         parseResult.issues.push(['warn', 'incomplete:range-request', msg])
       }
+      let chunker
+      if (!fileInfo.isGzipped) {
+        chunker = new ChunkedLineReader(file, ignoreLastLine)
+      } else {
+        chunker = new ChunkedLineReader(file, ignoreLastLine, true)
+      }
```
Comment (devonbush) [inline app/javascript/lib/validation/validate-file-content.js:413]:
```suggestion
      const chunker = new ChunkedLineReader(file, ignoreLastLine, fileInfo.isGzipped)
```

**Review by devonbush — APPROVED** [review]
Looks good.  Some style suggestions and a comment cleanup.

File: app/javascript/lib/validation/validate-file-content.js (hunk 4 above)
Comment (eweitz) [reply app/javascript/lib/validation/validate-file-content.js:419]: I vaguely recall trailing spaces in multi-line strings require a special solution with backticks, making them less an obvious choice here.

File: app/javascript/lib/validation/io.js (hunk 3 above)
Comment (eweitz) [reply app/javascript/lib/validation/io.js]: Yes, but I've refined it with some commentary to clarify why it's here.

**Comment by eweitz** [comment]
> As a future thought - the upload warning/error section is getting to be a little wordy when there's warning/s and error/s and I wonder if we should consider a UX-y update to make the most important bits stand out more

I like that idea, @ehanna4!  

As part of that, perhaps we could replace "Email us..." with a "Contact us" link that goes to a Zendesk form templated with fields relevant for problems adding data.  That'd help denoise the UI and also give us slightly better mechanisms to analyze what percent of support requests relate to adding data.

**Pull request merged**